### PR TITLE
Add DCA permissions

### LINKED ISSUE #6693: Missing permissions on image sizes
> So a user can not see an option in the list, but can duplicated said option and then have access to the new one. Not sure that's really intended?

It is not. We should fix this IMHO.

_Originally posted by @leofeyer in https://github.com/contao/contao/issues/6646#issuecomment-1870933846_

### PR DESCRIPTION
**Pull request by aschempp**
This superseeds https://github.com/contao/contao/pull/7591 and https://github.com/contao/contao/pull/6036. Implements https://github.com/contao/contao/issues/6693.

Would love to get some feedback @bezin @leofeyer 

## DCA access permissions
This PR adds a new `cud` configuration (taken from https://github.com/contao/contao/pull/7591), which allows to define permissions for each DCA table (that extends `DC_Table`).
image

It also allows DCAs to customize these permissions through `$GLOBALS['TL_DCA'][$table]['config']['permissions']`. For example
 - setting `$GLOBALS['TL_DCA'][$table]['config']['permissions'] = []` will remove the table, e.g. for `tl_job` or `tl_undo`
 - `terminal42/contao-leads` does not have a `create` permissions (https://github.com/terminal42/contao-leads/blob/main/contao/dca/tl_user_group.php). 
 - `terminal42/contao-node` has more than create/update/delete (https://github.com/terminal42/contao-node/blob/main/contao/dca/tl_user.php#L25). 

Both these cases are supported by the widget and voter.

## _Remote_ access to permission of elements

This PR dynamically adds a permission checkbox for records that have a respective user configuration. Inspired by https://github.com/contao/contao/pull/6036, it adds a checkbox wizards when editing records that have `$GLOBALS['TL_DCA'][$table]['config']['userRoot']` configured.

Bildschirmfoto 2025-11-18 um 09 58 22

_The permission field is currently not available in `edit-all` or `override-all` mode, but this could probably be improved upon._

## Automatic `list.sorting.root` config and update of access to new elements

Based on the `$GLOBALS['TL_DCA'][$table]['config']['userRoot']` configuration, the `list.sorting.root` config is automatically set for non-admin users. Currently, this code was duplicated for every DCA.
If a user adds a new element, the respective user/groups are also automatically given access to be allowed to edit that element. This code is currently duplicated in every DCA as well.

## TODO

- [x] The `cud` widget currently only exists for `tl_user_group` (similar to `alexf`), but probably needs to exist for `tl_user` as well (since the current `newsp` etc. exist for users as well)
- [x] Create/update/delete permissions are currently not checked automatically and need to be decided by individual voters.

### CODE REVIEW

**Comment by aschempp** [comment]
Failing ECS seems unrelated

File: core-bundle/src/Migration/Version507/FieldPermissionMigration.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+namespace Contao\CoreBundle\Migration\Version507;
+
+use Contao\CoreBundle\Migration\AbstractMigration;
+use Contao\CoreBundle\Migration\MigrationResult;
+use Contao\StringUtil;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Platforms\AbstractMySQLPlatform;
+use Doctrine\DBAL\Schema\AbstractSchemaManager;
+
+class FieldPermissionMigration extends AbstractMigration
```
Comment (Toflar) [inline core-bundle/src/Migration/Version507/FieldPermissionMigration.php]: We should have an `AbstractFieldPermissoinMigration` for re-using this to be consistent with the `AbstractBasicEntitiesMigration` concept. And this one for the `core-bundle` should then extend from it for the `formp` mapping.

File: core-bundle/tests/Command/MigrateCommandTest.php [hunk 2]
```diff
@@ -458,7 +458,6 @@ public function testAbortsOnWrongServerVersion(string $format): void
         if (interface_exists(ServerInfoAwareConnection::class)) {
             /** @phpstan-ignore class.notFound */
             $driverConnection = $this->createStub(ServerInfoAwareConnection::class);
-
```
Comment (Toflar) [inline core-bundle/tests/Command/MigrateCommandTest.php:461]: Unrelated?

**Review by Toflar — CHANGES_REQUESTED** [review]
We're still using `root` in an `onload_callback` at the end of the day which I think is a concept that we'll definitely need to replace in the future with some data provider concept within the `DC_Table` or so. However, I do see that this PR is a first step into this direction and it gets rid of a ton of repeating and tedious to maintain code.
So definitely 👍 on this one.

I don't even see the widget though?

Image

File: core-bundle/src/Migration/Version507/FieldPermissionMigration.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/Migration/Version507/FieldPermissionMigration.php]: Changed in 4dd7b25065344e21c257ae146283d0a5d70ddee8

**Comment by aschempp** [comment]
Failing Codecov is due to missing tests on migrations, but none of our migrations have tests and testing SQL commands makes little sense…

**Review by Toflar — COMMENTED** [review]
I'm trying the permissions a bit and will report here.

Issue with `tl_article`. When I disable `update`, I can still edit and update an article:

Image

Same for `tl_calendar`, so maybe the "update" (edit) permission is not checked correctly?

**Comment by aschempp** [comment]
Creating new calendars (and other elements) is fixed. I can't reproduce the article issue.

**Review by Toflar — DISMISSED** [review]
I think usability on the widget could be improved a bit with some JS that automatically updates the checkboxes because `create` without `update` makes no sense.
But that can also be improved later on.

File: core-bundle/contao/dca/tl_user.php [hunk 3]
```diff
@@ -444,7 +442,8 @@ public function handleUserProfile(DataContainer $dc)
 		$GLOBALS['TL_DCA'][$dc->table]['palettes'] = array
 		(
 			'__selector__' => $GLOBALS['TL_DCA'][$dc->table]['palettes']['__selector__'],
-			'default' => $GLOBALS['TL_DCA'][$dc->table]['palettes']['login']
+			'login' => $GLOBALS['TL_DCA'][$dc->table]['palettes']['login'],
```
Comment (leofeyer) [inline core-bundle/contao/dca/tl_user.php:445]: This seems strange to me. Why did you add it?

File: core-bundle/contao/languages/en/tl_user.xlf [hunk 4]
```diff
@@ -182,18 +182,39 @@
...
+      </trans-unit>
+      <trans-unit id="tl_user.cud.table_create">
+        <source>Create records in "%s"</source>
+      </trans-unit>
+      <trans-unit id="tl_user.cud.table_update">
+        <source>Update records of "%s"</source>
+      </trans-unit>
+      <trans-unit id="tl_user.cud.table_delete">
+        <source>Delete records of "%s"</source>
+      </trans-unit>
```
Comment (leofeyer) [inline core-bundle/contao/languages/en/tl_user.xlf:217]: Where are these used?

**Review by leofeyer — CHANGES_REQUESTED** [review]
There are two things that bother me:

1. I don't like that there is now a `BackendAccessVoter` and a `BackwardsCompatibilityBackendAccessVoter`. There should only be one per bundle which handles both the old fields and the new ones.

2. Being able to edit permissions per form / calendar / news archive etc. is very confusing and cumbersome. I'm talking about this:



<del>Previously, I could configure all of this in the user group at one central place. Now I have to navigate to every item and set the permissions separately.</del>

**Edit:** As clarified in the call, this is just a virtual field that shows all groups and the users which extend their group settings. So that's okay with me then 😊

**Comment by leofeyer** [comment]
As discussed in the call, we want to call them `LegacyBackendAccessVoter`.

File: core-bundle/contao/languages/en/tl_user.xlf (hunk 4 above)
Comment (aschempp) [reply core-bundle/contao/languages/en/tl_user.xlf:217]: This is for the invisible label of each checkbox

File: core-bundle/contao/dca/tl_user.php (hunk 3 above)
Comment (aschempp) [reply core-bundle/contao/dca/tl_user.php:445]: The current code would remove the login palette, which is necessary to check permissions on the user profile.

Comment (leofeyer) [reply core-bundle/contao/dca/tl_user.php:445]: I cannot reproduce this. If I revert the changes, I can still check edit my profile (tested with an admin user and a regular user).

File: core-bundle/config/services.yaml [hunk 5]
```diff
@@ -1181,8 +1181,21 @@ services:
             - '@security.access.decision_manager'
             - '@database_connection'
 
+    contao.security.data_container.user_profile_voter:
+        class: Contao\CoreBundle\Security\Voter\DataContainer\UserProfileVoter
+
+    contao.security.data_container_permission_voter:
+        class: Contao\CoreBundle\Security\Voter\DataContainer\DataContainerPermissionVoter
```
Comment (leofeyer) [inline core-bundle/config/services.yaml]: Can we rename the class like this?

```suggestion
    contao.security.data_container.table_permission_voter:
        class: Contao\CoreBundle\Security\Voter\DataContainer\TablePermissionVoter
```

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.
